### PR TITLE
Information schema

### PR DESCRIPTION
**Pull request by OllieJones**
#19 #20 #21 #22 

Escaping LIKE literal % and _.
More information_schema and metadata compatiblity.
Unique constraint violations, and other errors when running queries, now detected and reported.
SUBSTRING and CHAR_LENGTH now handled by rewriting the query to SUBSTR and LENGTH.

### CODE REVIEW

File: tests/WP_SQLite_Metadata_Tests.php [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+					return str_ends_with( $row->Name, 'comments' );
+				}
+			)
+		);
+		$this->assertEquals( 'wp_comments', $rows[0]->Name );
+		$this->assertEquals( 4, $rows[0]->Rows );
+	}
+
+	private function assertTableEmpty( $table_name, $empty ) {
```
Comment (adamziel) [inline tests/WP_SQLite_Metadata_Tests.php:272]: ❤️

File: tests/WP_SQLite_Query_Tests.php [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+
+		$this->assertQuery( $query );
+		$actual = $this->engine->get_query_results();
+		$this->assertEquals( 0, count( $actual ) );
+	}
+
+	public function testCharLengthComplex() {
+		$query = <<<'QUERY'
+SELECT option_name,
+       CHAR_LENGTH(
```
Comment (adamziel) [inline tests/WP_SQLite_Query_Tests.php:281]: It's so nice that complex queries like this one work

File: tests/WP_SQLite_Query_Tests.php [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+				CASE WHEN option_name LIKE '\_site\_transient\_%'
+				     THEN '_site_transient_'
+                     WHEN option_name LIKE '\_transient\_%'
```
[the remark this answers is not in the document]
Comment (aristath) [reply tests/WP_SQLite_Query_Tests.php:284]: I was planning to do a PR _after_ this one gets merged to fix various formatting issues and optimize some performance things I've seen, so I wouldn't worry too much about it right now... I'll be doing a follow-up PR to fix any and all issues like that

File: wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php [hunk 4]
```diff
@@ -587,7 +575,7 @@ public function least() {
 	public function greatest() {
 		$arg_list = func_get_args();
 
-		return "max($arg_list)";
+		return max( $arg_list );
```
[the remark this answers is not in the document]
Comment (adamziel) [reply wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php:578]: Did that even work before? This part of the codebase carried over from the regexp-based plugin and I find it surprising that returning an SQL expression would actually execute it.

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 5]
```diff
@@ -272,6 +277,48 @@ class WP_SQLite_Translator {
...
+	 * @var bool
+	 */
+	private $has_group_by = false;
+
+	/**
+	 * 0 if no LIKE is in progress, otherwise counts nested parentheses.
+	 *
+	 * @var int
+	 */
+	private $like_expression_nesting = 0;
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-translator.php:306]: A generic stack of expression would scale better. There's already a `call_stack` in WP_SQLite_Query_Rewriter and I worry we may eventually end up with many different ways of tracking nesting level (like  `like_expression_nesting`, `case_expression_nesting` etc). Not a blocker here, but let's keep an eye on how this code evolves.

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 6]
```diff
@@ -476,6 +535,36 @@ private function prepare_directory() {
 	 */
 	public function query( $statement, $mode = PDO::FETCH_OBJ, ...$fetch_mode_args ) { // phpcs:ignore WordPress.DB.RestrictedClasses
 		$this->flush();
+		if ( function_exists( 'apply_filters' ) ) {
+			/**
+			 * Filters queries before they are translated and run.
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-translator.php:540]: ❤️

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 7]
```diff
@@ -1577,6 +1671,29 @@ private function preprocess_string_literal( $value ) {
 		return $value;
 	}
 
+	private function preprocess_like_expr( &$token ) {
+		/*
+		 * This code handles escaped wildcards in LIKE clauses.
+		 * If we are within a LIKE experession, we look for \_ and \%, the
+		 * escaped LIKE wildcards, the ones where we want a literal, not a
+		 * wildcard match. We change the \ escape for a ~ character,
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-translator.php:1679]: Is `~` a valid escape character in SQLite?

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 8]
```diff
@@ -2076,6 +2233,178 @@ private function translate_regexp_functions( $token ) {
...
+			)
+		) {
+			return false;
+		}
+		$next = $this->rewriter->peek_nth( 2 )->value;
+		if ( 'BY' !== strtoupper( $next ) ) {
+			return false;
+		}
+
+		$this->has_group_by = true;
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-translator.php:2258]: Would `$this->last_reserved_keyword` or something to that effect work? This here is only meant to have impact on the very next `HAVING` clause, but in this current form a nested query that has a `GROUP BY` would trigger a false-positive for the outer query that has an ungrouped `HAVING` query.

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 9]
```diff
@@ -2076,6 +2233,178 @@ private function translate_regexp_functions( $token ) {
...
+	 *
+	 * @return bool
+	 */
+	private function translate_like_escape( $token ) {
+
+		if ( 0 === $this->like_expression_nesting ) {
+			$match = $token->matches( WP_SQLite_Token::TYPE_KEYWORD, null, array( 'LIKE' ) );
+			/* is this the LIKE keyword? If so set the flag. */
+			if ( $match ) {
+				$this->like_expression_nesting = 1;
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-translator.php:2305]: Or `$this->like_expression_nesting ++;` for consistency?

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 10]
```diff
@@ -2076,6 +2233,178 @@ private function translate_regexp_functions( $token ) {
...
+			}
+		} else {
+			/* open parenthesis during LIKE parameter, count it. */
+			if ( $token->matches( WP_SQLite_Token::TYPE_OPERATOR, null, array( '(' ) ) ) {
+				$this->like_expression_nesting ++;
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-translator.php:2310]: Oh so this counts parenthesis nesting! `$this->rewriter->depth` would do the same trick, you'd just have to store the depth of the initial `LIKE` – OR make sure it's pushed to the call stack.

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 11]
```diff
@@ -2076,6 +2233,178 @@ private function translate_regexp_functions( $token ) {
...
+			        || $token->matches( WP_SQLite_Token::TYPE_DELIMITER, null, array( ';' ) ) || ( WP_SQLite_Token::TYPE_DELIMITER === $token->type && null === $token->value )
+			        || $token->matches( WP_SQLite_Token::TYPE_OPERATOR, null, array( ')', ',' ) );
+
+			if ( $match ) {
+				/* Here we have another keyword encountered with the LIKE in progress.
+				 * Emit the ESCAPE clause.
+				 */
+				if ( $this->like_escape_count > 0 ) {
+					/* If we need the ESCAPE clause emit it. */
+				$this->rewriter->add( new WP_SQLite_Token( ' ', WP_SQLite_Token::TYPE_DELIMITER ) );
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-translator.php:2333]: nit: formatting

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 12]
```diff
@@ -2076,6 +2233,178 @@ private function translate_regexp_functions( $token ) {
...
+
+			/* close parenthesis matching open parenthesis during LIKE parameter, count it. */
+			if ( $this->like_expression_nesting > 1 && $token->matches( WP_SQLite_Token::TYPE_OPERATOR, null, array( ')' ) ) ) {
+				$this->like_expression_nesting --;
+
+				return false;
+			}
+
+			/* a keyword, a commo, a semicolon, the end of the statement, or a close parenthesis */
+			$match = $token->matches( WP_SQLite_Token::TYPE_KEYWORD )
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-translator.php:2323]: A different variable name would make the logic more obvious:

```suggestion
			$is_like_finished = $token->matches( WP_SQLite_Token::TYPE_KEYWORD )
```

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 13]
```diff
@@ -2076,6 +2233,178 @@ private function translate_regexp_functions( $token ) {
...
+	 * If the flag is set, we emit ESCAPE '\' before the next keyword.
+	 *
+	 * @param WP_SQLite_Token $token The token to translate.
+	 *
+	 * @return bool
+	 */
+	private function translate_like_escape( $token ) {
+
+		if ( 0 === $this->like_expression_nesting ) {
+			$match = $token->matches( WP_SQLite_Token::TYPE_KEYWORD, null, array( 'LIKE' ) );
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-translator.php:2302]: A different variable name would make the logic clearer:
```suggestion
			$is_like_keyword = $token->matches( WP_SQLite_Token::TYPE_KEYWORD, null, array( 'LIKE' ) );
```

File: wp-includes/sqlite/class-wp-sqlite-translator.php [hunk 14]
```diff
@@ -1577,6 +1671,29 @@ private function preprocess_string_literal( $value ) {
...
+		 * These \_ and \% escape sequences are in the token name, because
+		 * the lexer has already done stripcslashes on the value.
+		 */
+		if ( $this->like_expression_nesting > 0 ) {
+			/* Remove the quotes around the name. */
+			$unescaped_value = mb_substr( $token->token, 1, -1, 'UTF-8' );
+			if ( str_contains( $unescaped_value, '\_') || str_contains( $unescaped_value, '\%') ) {
+				$this->like_escape_count ++;
+				return str_replace(
+					array ( '\_', '\%' ),
```
[the remark this answers is not in the document]
Comment (adamziel) [reply wp-includes/sqlite/class-wp-sqlite-translator.php:1690]: Yeah, a double backslash turns `_` into a wildcard again:
```sql
select 1 WHERE '\\ot' LIKE '\\_%' ESCAPE '\';
```

**Comment by adamziel** [comment]
Great work on this one @OllieJones! I left some notes but this is the right direction. Do core unit tests still pass?

**Comment by OllieJones** [comment]
By "core" unit tests I guess you mean the ones with this plugin. They do indeed pass. 

I've added features and unit tests based on stuff I found in some data-maint plugins. 

Thanks for your review; I'll incorporate your changes when I get a chance. `LIKE '\_'` and `LIKE '\%'` in my code are a nasty kludge.

File: wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (OllieJones) [reply wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php:578]: LEAST and GREATEST didn't work before I made the changes.

**Comment by adamziel** [comment]
> By "core" unit tests I guess you mean the ones with this plugin. They do indeed pass.

Oh I meant WordPress core with this plugin installed. I've got [almost all of them to pass](https://github.com/WordPress/sqlite-database-integration/pull/9#issuecomment-1437018625) in #9 and it would be great to avoid regressions.

**Comment by aristath** [comment]
I think we can merge this one as-is and continue iterating in follow-up PR.
What do you think @adamziel @OllieJones ?

**Comment by adamziel** [comment]
@aristath I'd still love to understand how it impacts WordPress core tests. If it doesn't break any then yes, let's merge!

File: wp-includes/sqlite/class-wp-sqlite-translator.php (hunk 5 above)
Comment (OllieJones) [reply wp-includes/sqlite/class-wp-sqlite-translator.php:306]: Added a @todo.

File: wp-includes/sqlite/class-wp-sqlite-translator.php (hunk 7 above)
Comment (OllieJones) [reply wp-includes/sqlite/class-wp-sqlite-translator.php:1679]: Bogus comment, now fixed. I'm using the ASCII SUB (substitute, 0x1a) character. It tests out fine.

File: wp-includes/sqlite/class-wp-sqlite-translator.php (hunk 8 above)
Comment (OllieJones) [reply wp-includes/sqlite/class-wp-sqlite-translator.php:2258]: Man, that's an edge case. Some plugin author's writing SQL code with subqueries AND misuse of HAVING, they can doggone well fix it. The testcase for this came from Advanced Database Cleaner, without subqueries. I'll contact the author.

**Comment by OllieJones** [comment]
Fixes from your reviews now pushed.

**Comment by aristath** [comment]
Merged the PR, then fixed some CS issues in https://github.com/WordPress/sqlite-database-integration/commit/65c50f475bceccca67c51b56f142beb26c289d94 and backported the changes to https://github.com/WordPress/wordpress-develop/pull/3220 as well 👍

File: wp-includes/sqlite/class-wp-sqlite-translator.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (adamziel) [reply wp-includes/sqlite/class-wp-sqlite-translator.php:1691]: @OllieJones This was marked as resolved but the problem is still there.
